presto-main/src/main/java/com/facebook/presto/operator/MergeHashSort.java

```diff
@@ -0,0 +1,197 @@
+        for (int i = 0; i < keyTypes.size(); i++) {
+            hashChannels[i] = i;
+        }
+        HashGenerator hashGenerator = new InterpretedHashGenerator(keyTypes, hashChannels);
+
+        return new PageRewriteIterator(
+                hashGenerator,
+                allTypes,
+                Iterators.mergeSorted(
+                        channelIterators,
+                        (PagePosition left, PagePosition right) -> comparePages(hashGenerator, left, right)));
+    }
+
+    private static int comparePages(HashGenerator hashGenerator, PagePosition left, PagePosition right)
+    {
+        if (left.isPositionOutOfPage() && right.isPositionOutOfPage()) {
+            return 0;
+        }
+        if (left.isPositionOutOfPage()) {
+            return -1;
+        }
+        if (right.isPositionOutOfPage()) {
+            return 1;
+        }
+
+        long leftHash = hashGenerator.hashPosition(left.getPosition(), left.getPage());
+        long rightHash = hashGenerator.hashPosition(right.getPosition(), right.getPage());
+
+        return Long.compare(leftHash, rightHash);
+    }
+
+    public static class PagePosition
+    {
+        private final Page page;
+        private final int position;
+
+        public PagePosition(Page page, int position)
+        {
+            this.page = requireNonNull(page, "page is null");
+            this.position = requireNonNull(position, "position is null");
+        }
+
+        public Page getPage()
+        {
+            return page;
+        }
+
+        public int getPosition()
+        {
+            return position;
+        }
+
+        public boolean isPositionOutOfPage()
+        {
+            return position >= page.getPositionCount();
+        }
+    }
+
+    public static interface PagePositions extends Iterator<PagePosition>
+    {
+    }
+
+    public static class SingleChannelPagePositions
+            implements PagePositions
+    {
+        private final Iterator<Page> channel;
+        private PagePosition current;
+
+        public SingleChannelPagePositions(Iterator<Page> channel)
+        {
+            this.channel = requireNonNull(channel, "channel is null");
+        }
+
+        @Override
+        public boolean hasNext()
+        {
+            return channel.hasNext() || (current != null && current.getPosition() + 1 < current.getPage().getPositionCount());
+        }
+
+        @Override
+        public PagePosition next()
+        {
+            if (current == null || current.getPosition() + 1 >= current.getPage().getPositionCount()) {
+                current = new PagePosition(channel.next(), 0);
+            }
+            else {
+                current = new PagePosition(current.getPage(), current.getPosition() + 1);
+            }
+            return current;
+        }
+    }
+
+    /**
+     * This class rewrites iterator over PagePosition to iterator over Pages.
+     * Positions with same hash value are guaranteed to be in the same result page.
```

Comment:
We need it to simplify merging and building final result. This let us assume while reading spills, that every `Page` we get from `MergeHashSort` contains all relevant entries to produce final output. Otherwise, we would have to manually validate whether next `Page` contains positions new keys, or keys that we are already aggregated. Further more, those keys could change in the middle of the `Page`. We could end up with situation that we don't have memory for adding new keys to the `HashAggregationBuilder` but next Page would contain both old keys that we need to release memory and new keys (for which to aggregate we don't have memory).

Secondly, this is shouldn't be big issue:
1. those Pages are immediately consumed and aggregated
2. there shouldn't be that many keys with the same value, because the more keys with same value the less memory is actually needed to perform aggregation, so if we are already spilling that means that we have lots of unique aggregation keys. I would expect this code to only marginally exceed maximum size.
3. Those are already partially aggregated Pages

I would postpone fixing this until we see that this could be an issue, or at least as a separate follow up improvement.